lib/rules/one-var.js

```diff
@@ -249,109 +246,151 @@ module.exports = {
             return true;
         }
 
+        /**
+         * Checks a given VariableDeclaration node for errors.
+         * @param {ASTNode} node The VariableDeclaration node to check
+         * @returns {void}
+         * @private
+         */
+        function checkVariableDeclaration(node) {
+            const parent = node.parent;
+            const type = node.kind;
 
-        //--------------------------------------------------------------------------
-        // Public API
-        //--------------------------------------------------------------------------
+            if (!options[type]) {
+                return;
+            }
 
-        return {
-            Program: startFunction,
-            FunctionDeclaration: startFunction,
-            FunctionExpression: startFunction,
-            ArrowFunctionExpression: startFunction,
-            BlockStatement: startBlock,
-            ForStatement: startBlock,
-            ForInStatement: startBlock,
-            ForOfStatement: startBlock,
-            SwitchStatement: startBlock,
+            const declarations = node.declarations;
+            const declarationCounts = countDeclarations(declarations);
 
-            VariableDeclaration(node) {
-                const parent = node.parent;
-                const type = node.kind;
+            // consecutive
+            const nodeIndex = (parent.body && parent.body.length > 0 && parent.body.indexOf(node)) || 0;
 
-                if (!options[type]) {
-                    return;
-                }
+            if (nodeIndex > 0) {
```

Comment:
Have you considered making this a `BlockStatement` (and `Program` and `SwitchCase`) listener rather than a `VariableDeclaration` listener? It seems like there's a significant amount of code going into getting the previous statement for a given variable declaration, so I'm wondering if it would be easier to do a single pass over the statements in a `BlockStatement` instead and apply an operation on the `VariableDeclaration` nodes (while keeping track of the previous statement at any given point).